File: LC/[0153] Find Minimum in Rotated Sorted Array.cpp
class Solution {
public:
    int findMin(vector<int>& nums) {
        int ll=0;
        int rr=nums.size()-1;
        int ans=-1;
        int N=nums.size();
        if(nums[0] <= nums[N-1]) return nums[0];

        while(ll<=rr){
            int mm=ll+(rr-ll)/2;
            if(nums[mm] > nums[N-1]){ //on the left, need to search right.  left cannot have answer
                ll=mm+1;
            }else{
                ans=nums[mm]; //hold the potential answer
                rr=mm-1;
            }
        }
        return ans;
    }
};

File: LC/[0153] Find Minimum in Rotated Sorted Array.py
#############
# 20231125
#############
class Solution:
    def findMin(self, nums: List[int]) -> int:
        N = len(nums)
        ll, rr, ans = 0, N-1, -1
        if nums[0] < nums[-1]:
            return nums[0]

        while ll <= rr:
            mm = ll+(rr-ll)//2

            if nums[mm] < nums[0]:  # on the right
                ans = mm
                rr = mm-1
            else:  # on the left
                ll = mm+1

        return nums[ans]


##############################
class Solution:
    def findMin(self, nums: List[int]) -> int:
        ll, rr, ans = 0, len(nums)-1, -1

        if nums[0] <= nums[-1]:
            return nums[0]

        while ll <= rr:
            mm = ll+(rr-ll)//2

            if nums[mm] >= nums[-1]:  # on the left
                ll = mm+1
            else:  # on the right
                ans = mm
                rr = mm-1

        return nums[ans]
